### PR TITLE
{citra,yuzuPackages}: remove package

### PR DESCRIPTION
**Pull request by shyim**
## Description of changes

Yuzu has been taken down because of Nintendo.
image

The Git repos are already gone https://github.com/yuzu-emu/yuzu

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by AndersonTorres — APPROVED** [review]
In my heart I am disapproving this.

**Comment by shyim** [comment]
https://github.com/pineappleEA/pineapple-src is still up, but for how long 😓

**Comment by SuperSamus** [comment]
https://github.com/citra-emu/citra is down too.

**Comment by shyim** [comment]
Citra can be done in a second PR? 🤔

**Comment by AndersonTorres** [comment]
Citra  was taken down too :/

**Comment by SuperSamus** [comment]
I'm not a lawyer, but I imagine that they need to be purged off from Hydra too?

**Comment by AndersonTorres** [comment]
Better safe than sorry.

**Comment by uninsane** [comment]
Software Heritage has up-to-date archives if you didn't get them in time, and there are out-of-tree _yuzu flake_'s on other forges for maintaining the nix expressions.

**Comment by shyim** [comment]
Removed now Citra too 😭

File: pkgs/top-level/aliases.nix [hunk 1]
```diff
@@ -1220,10 +1222,11 @@ mapAliases ({
   yafaray-core = libyafaray; # Added 2022-09-23
   yarn2nix-moretea-openssl_1_1 = throw "'yarn2nix-moretea-openssl_1_1' has been removed."; # Added 2023-02-04
   yubikey-manager4 = throw "yubikey-manager4 has been removed, since it is no longer required by yubikey-manager-qt. Please update to yubikey-manager."; # Added 2024-01-14
-  yuzu-ea = yuzuPackages.early-access; # Added 2022-08-18
-  yuzu-early-access = yuzuPackages.early-access; # Added 2023-12-29
-  yuzu = yuzuPackages.mainline; # Added 2021-01-25
-  yuzu-mainline = yuzuPackages.mainline; # Added 2023-12-29
+  yuzu-ea = throw "yuzu-ea has been removed from nixpkgs, as it has been deleted upstream"; # Added 2024-03-04
```
Comment (AndersonTorres) [inline pkgs/top-level/aliases.nix:1225]: I would wrote "taken down" instead of "deleted". 
But it is just me angry and sad.

**Review by AndersonTorres — APPROVED** [review]

File: pkgs/top-level/aliases.nix (hunk 1 above)
Comment (shyim) [reply pkgs/top-level/aliases.nix]: updated, I was not sure what to write 😞

**Comment by niklaskorz** [comment]
This feels a bit premature. There was no ruling that Yuzu is illegal. The out of court settlement, from my understanding, is mostly based on the situation that the company developing Yuzu took money for early access releases that specifically supported a yet unreleased game and also did not forbid discussion of piracy on their private Discord. I feel like for now it may make more sense to point this to a fork that contains the commits references from nixpkgs, and maybe switch it to a more active community maintained one once such is established. As Yuzu itself is GPL licensed, it will not disappear anytime soon.

**Comment by AndersonTorres** [comment]
Right now our expression does compile nothing. Waiting for an "oficial fork" does not make much sense.

On the other hand, there are copies of it floating around the forges currently.

**Comment by thiagokokada** [comment]
> Right now our expression does compile nothing. Waiting for an "oficial fork" does not make much sense.
> 
> On the other hand, there are copies of it floating around the forges currently.

AFAIK, Hydra caches forever and a `src` is also a derivation, so we do have a copy and this will work essentially forever.

**Comment by thiagokokada** [comment]
> This feels a bit premature. There was no ruling that Yuzu is illegal. The out of court settlement, from my understanding, is mostly based on the situation that the company developing Yuzu took money for early access releases that specifically supported a yet unreleased game and also did not forbid discussion of piracy on their private Discord. I feel like for now it may make more sense to point this to a fork that contains the commits references from nixpkgs, and maybe switch it to a more active community maintained one once such is established. As Yuzu itself is GPL licensed, it will not disappear anytime soon.

IMO, the repository will not come back. If there is a maintained fork that gets traction, we can introduce it to `nixpkgs`, but for now it doesn't make sense to keep it regardless of the court settlement.

Or for an example, https://github.com/libretro/citra still exists, so it doesn't make sense to remove `libretro.citra` because upstream still exist. But for `yuzu` and `citra` themselves, the upstream is dead so we should also delete and deprecate the packages.

File: pkgs/applications/emulators/retroarch/cores.nix [hunk 2]
```diff
@@ -317,27 +317,6 @@ in
...
-      "BUILD_DATE=01/01/1970_00:00"
-    ];
-    # FIXME: build fail with GCC13:
-    # error: 'mic_device_name' has incomplete type
-    stdenv = gcc12Stdenv;
-    meta = {
-      description = "Port of Citra to libretro";
-      license = lib.licenses.gpl2Plus;
-    };
-  };
```
Comment (thiagokokada) [inline pkgs/applications/emulators/retroarch/cores.nix:339]: As I said before, https://github.com/libretro/citra still exists so I don't see a reason for removal.

Comment (thiagokokada) [reply pkgs/applications/emulators/retroarch/cores.nix:339]: Regardless of the decision here, can you move this deletion to another commit? Something like:

```
libretro.citra: remove package
```

**Comment by Scrumplex** [comment]
I am not a lawyer.

As others have already mentioned, yuzu is not illegal. The code is GPL-3.0-or-later and even if Nintendo owns the copyright to the code now, they can't retroactively revoke the prior license, which means it should still be freely redistributable.

File: pkgs/applications/emulators/yuzu/early-access/default.nix [hunk 3]
```diff
@@ -1,37 +0,0 @@
...
-# but the Windows distributions include a source tarball, which in turn
-# includes the full git metadata. So, grab that and rehydrate it.
-# This has the unfortunate side effect of requiring two FODs, one
-# for the Windows download and one for the full repo with submodules.
-let
-  sources = import ./sources.nix;
-
-  zip = fetchzip {
-    name = "yuzu-ea-windows-dist";
-    url = "https://github.com/pineappleEA/pineapple-src/releases/download/EA-${sources.version}/Windows-Yuzu-EA-${sources.version}.zip";
```
Comment (niklaskorz) [inline pkgs/applications/emulators/yuzu/early-access/default.nix:12]: The repository is still online, so no need to remove this package: https://github.com/pineappleEA/pineapple-src

**Comment by SuperSamus** [comment]
> the company developing Yuzu took money for early access releases that specifically supported a yet unreleased game

[This is false](https://reddit.idevicehacked.com/r/yuzu/comments/1b51wuf/yuzu_did_not_play_totk_before_the_release_date/), PRs and support for unreleased games were denied.

> did not forbid discussion of piracy on their private Discord

False, they even had a bot for that.
image

> Hydra caches forever

[Not necessarily](https://discourse.nixos.org/t/upcoming-garbage-collection-for-cache-nixos-org/39078).

> this will work essentially forever

Nixpkgs in theory should be able to build everything even without Hydra.
Now, it is true that (as the link above says) Hydra will not delete source codes due to the fact that old project sometimes disappear, but I think it is meant in the case someone checks out an old version of nixpkgs, not for making it possible to leave a dead package in nixpkgs' HEAD.
This is me speculating, though, so I'll make no further comments.

> There was no ruling that Yuzu is illegal

It is true that [Exhibit A](https://storage.courtlistener.com/recap/gov.uscourts.rid.56980/gov.uscourts.rid.56980.10.1.pdf) specifically mentions "[the] Defendant enjoining it and its members" and "third parties acting in active
concert and participation with Defendant", and... I admit I'm not qualified for further comments.

And Citra was caught in the crossfire simply because they shared developers, Nintendo didn't say anything about it specifically.

**Comment by thiagokokada** [comment]
> Now, it is true that (as the link above says) Hydra will not delete source codes due to the fact that old project sometimes disappear, but I think it is meant in the case someone checks out an old version of nixpkgs, not for making it possible to leave a dead package in nixpkgs' HEAD.

AFAIK this decision is by design, and not only for checkouts of older versions of nixpkgs, but because this is the only way to ensure reproducibility. I can't find the source for this though, so I may as well be misremembering.

Anyway, I support the decision to delete any version of `citra`/`yuzu` that has no upstream. The fact that they have no upstream can causes a bunch of other issues that is not related to code availability, e.g.: who is going to fix build issues once we bump `glibc`, `gcc` or whatever else? Without the source code available upstream this task would be difficult (of course, not impossible, but it is more work than it should be).

Also, while the decision may be premature, I also don't want to add any of the forks that are appearing right now since they may or may not gather enough popularity. Once we have a clear winner I think we can reintroduce it without much issues. The code for the derivations will be preserved forever in Git anyway, so it should be easier to grab the old diffs and restore.

**Comment by JoshuaFern** [comment]
I support the removal and I don't see it as premature, the upstream projects were deleted from Github.

It's not being removed from nixpkgs because it's "illegal software", it's just now unmaintained with no upstream.

After the dust settles we'll probably see forks, and there's still plenty of ways to use Yuzu and Citra in the mean time.

File: pkgs/applications/emulators/retroarch/cores.nix (hunk 2 above)
Comment (shyim) [reply pkgs/applications/emulators/retroarch/cores.nix]: I restored that. As you the repo is up :)

**Comment by shyim** [comment]
I restored the Retroarch Citra :)

**Review by thiagokokada — APPROVED** [review]

**Comment by leiserfg** [comment]
Can't we remove mainline only and keep EA that is fetched from another repo? If it starts failing to compile at some point and there is no new upstream, then remove it.

**Comment by thiagokokada** [comment]
### 

> Can't we remove mainline only and keep EA that is fetched from another repo? If it starts failing to compile at some point and there is no new upstream, then remove it.

My main issue with it is that the name itself is confusing: it should have been called `pineapple-ea` or something. Calling it `yuzu-early-access` does look like the upstream is still `yuzu` itself.

This is also why I think we need to wait until the situation stabilises: it may as well that pineappleEA becomes the new upstream, but I think this is still too soon to say.

**Comment by thiagokokada** [comment]
> Can't we remove mainline only and keep EA that is fetched from another repo? If it starts failing to compile at some point and there is no new upstream, then remove it.  

Also, I am less concerned about the source code itself being available (as pointed above, Hydra caches FOD forever), and more about if we have an upstream giving the code proper support.

**Comment by JoshuaFern** [comment]
> Can't we remove mainline only and keep EA that is fetched from another repo? If it starts failing to compile at some point and there is no new upstream, then remove it.

pineapple-ea just publishes GPL'ed Yuzu EA source code. Even if their repository stays online it's not a fork, it's more like a mirror, and the code is still unmaintained, there's no upstream, etc..

I think it's best to remove for now and see what happens in the future.

**Review by ghost — APPROVED** [review]

**Comment by RGBCube** [comment]
Could it be re-added under suyu https://gitlab.com/suyu-emu/suyu ?

**Comment by AndersonTorres** [comment]
> Could it be re-added under suyu https://gitlab.com/suyu-emu/suyu ?

Nope.

> Status
> We are trying to get the builds working. We are in need of developers. Join our discord to contribute.

Let's wait they gain more traction.

**Comment by Atemu** [comment]
> I support the decision to delete any version of `citra`/`yuzu` that has no upstream. The fact that they have no upstream can causes a bunch of other issues that is not related to code availability, e.g.: who is going to fix build issues once we bump `glibc`, `gcc` or whatever else? Without the source code available upstream this task would be difficult (of course, not impossible, but it is more work than it should be).

I fully agree that this is a problem.

What I do not agree on is that a package must be removed due to such a problem. For the time being, Yuzu builds for anyone who can access cache.nixos.org which is practically everyone.

Besides, we have lots of packages where the source isn't available anymore (or never was, proprietary software anyone?).  
We should strive towards having available sources (that's why we have the mirrors infrastructure) and not doing so warrants potentially marking a package as problematic but *removal* is a complete overreaction IMHO.

Some of our packages that have *far* worse issues than the upstream source having died that are still in the tree and don't hurt anyone.

> Also, while the decision may be premature, I also don't want to add any of the forks that are appearing right now since they may or may not gather enough popularity. Once we have a clear winner I think we can reintroduce it without much issues. The code for the derivations will be preserved forever in Git anyway, so it should be easier to grab the old diffs and restore.

I'm similarly cautious but it doesn't really matter as we don't have to trust the forks for this purpose specifically. Because they contain all the previous revisions, we'd treat them as a mirror and simply lock the srcs to the specific revisions they would have pointed at before. It'd just be a different URL that serves the same content; hash would stay the same.

***

I'd certainly not have this reach the stable channel (absolutely no good reason for the breakage) and will propose to revert this.

How to handle "the Yuzu issue" requires a bit more discussion still and I see no point in removing it this soon already because it might potentially cause minor issues *years* down the line.  
Once we reach the point where Yuzu causes *actual* issues affecting real users, *then* we can start thinking about removing it. Though even then marking the package with a harsher warning or requiring its source to be manually added to the Nix store would be the appropriate action, not removal.  
At that point, we'll likely have an established continuation of the project too.

**Comment by thiagokokada** [comment]
> Besides, we have lots of packages where the source isn't available anymore (or never was, proprietary software anyone?).

For packages that are open source and the source is not available anymore, we should strive to remove them. Unmaintained software is a maintenance burden and we already have way too many packages in nixpkgs, IMO.

For proprietary software, AFAIK we generally have either of those 2 approaches:
- We have a derivation that uses `requireFile` and let the user download the software themselves
- The derivation itself downloads the source if possible

The first approach is really difficult to check if the source is still available, while the second one AFAIK people generally works to keep it up-to-date or at least available, so I don't see much of the issue here.

**Comment by Atemu** [comment]
> The first approach is really difficult to check if the source is still available, while the second one AFAIK people generally works to keep it up-to-date or at least available, so I don't see much of the issue here.

Yes and that's exactly my point. If it's not an issue for other packages, why would it be an issue for Yuzu?

**Comment by thiagokokada** [comment]
> Yes and that's exactly my point. If it's not an issue for other packages, why would it be an issue for Yuzu?

I am arguing it is an issue for other packages, it is just more difficult to track.

Yuzu/Citra kinda was really easy to reason its removal because it was in news everywhere, but most packages will disappear in the world and nobody will blink an eye. But my argument here is that if we find those packages we also need to remove them from nixpkgs.

**Comment by AndersonTorres** [comment]
> What I do not agree on is that a package must be removed due to such a problem. For the time being, Yuzu builds for anyone who can access cache.nixos.org which is practically everyone.
> 

1. Nix allows pinning a rev of Nixpkgs. I see no problem here.
2. Nix is, primarily, a source-based package manager. It makes few to no sense to keep a buildscript for a source that does not exist anymore.
3. You yourself gave the example of packages that can't be cached. Then we have  exceptions to deal.

> Besides, we have lots of packages where the source isn't available anymore (or never was, proprietary software anyone?). We should strive towards having available sources (that's why we have the mirrors infrastructure) and not doing so warrants potentially marking a package as problematic but _removal_ is a complete overreaction IMHO.
> 

I don't think so. The upstream does not exist anymore, and keeping an expression for it makes few to no sense.
As an historical curiousity (fulfilled by the fact it is a VCS with (a huge) history), maybe, but as a build script? Nope.

> Some of our packages that have _far_ worse issues than the upstream source having died that are still in the tree and don't hurt anyone.

Hey, this is an argument for cleaning up the codebase, not an argument for keeping it bit-rotting!

**Comment by Atemu** [comment]
> my argument here is that if we find those packages we also need to remove them from nixpkgs.

No. We do not as a general rule need to remove building, working packages where the upstream source happened to have gone MIA. Especially not if a substitute is likely to come up within a month or so and doubly enjoyed so if it's used by thousands.

We should create issues on them, ping maintainers, mark them as problematic etc. because a fetchable upstream is obviously something we want but removal without notice is not an appropriate action for any such package.

> 1. Nix allows pinning a rev of Nixpkgs. I see no problem here.

You cannot use graphical applications from revisions other than your system revision without terrible hacks. This is not an option for users and you should not be recommending it without mentioning that major gotcha.

> Nix is, primarily, a source-based package manager. It makes few to no sense to keep a buildscript for a source that does not exist anymore.

For anyone who can use cache.nixos.org, it does indeed exist. If that went away too, then I'd start considering touching the package at all. Even in that case removal without notice is not the appropriate action for a widely used package but marking the source as broken and/or looking for a mirror.

> You yourself gave the example of packages that can't be cached. Then we have exceptions to deal.

I do not understand what this sentence is trying to say.

> The upstream does not exist anymore, and keeping an expression for it makes few to no sense.

Even if this was valid (which it is not, the src is cached), this only concerns the src, not the rest of the drv. While others are preferable of course, `requireFile` is absolutely a valid way to declare a source. This is really not much different.

> Hey, this is an argument for cleaning up the codebase, not an argument for keeping it bit-rotting!

I have yet to see a good argument for how this "cleaning up" is going to help anyone.

Actually dead code that does not build anymore, where the package has no responsive maintainers and/or causes a concrete issue, yeah remove that shit.  
Code that still works (I just built it), has maintainers and does not cause any *concrete* issue whatsoever for anyone? Why in the world would you remove that? Seriously.

**Comment by leiserfg** [comment]
> No. We do not as a general rule need to remove building, working packages where the upstream source happened to have gone MIA. Especially not if a substitute is likely to come up within a month or so and doubly enjoyed so if it's used by thousands.

Thankfully it does not happen very often, but that kind of "what to do" should be documented somewhere to avoid this kind of situations.

**Comment by AndersonTorres** [comment]
> Thankfully it does not happen very often, but that kind of "what to do" should be documented somewhere to avoid this kind of situations.

The previous occurrence was way worse: https://github.com/NixOS/nixpkgs/pull/157959

And at least in the previous case we had a still-alive source code.

**Comment by thiagokokada** [comment]
I will not lock this thread (or at least yet) but can we please keep the discussion in #295587 instead?

For one this PR is already merged, for another having the same discussion happening in 2 different places is confusing.

**Pull request merged**